pkg/kubelet/cpumanager/cpu_manager.go

```diff
@@ -0,0 +1,363 @@
+)
+
+type kletGetter interface {
+	GetPods() []*v1.Pod
+	GetCachedMachineInfo() (*cadvisorapi.MachineInfo, error)
+}
+
+type runtimeService interface {
+	UpdateContainerResources(id string, resources *runtimeapi.LinuxContainerResources) error
+}
+
+type policyName string
+
+// Manager interface provides methods for kubelet to manage pod cpus
+type Manager interface {
+	Start()
+
+	// RegisterContainer registers a container with the cpuset manager
+	// resulting in synchronous setting on the container cpuset.  This
+	// is called after RegisterContainer(), which creates the containerID, and
+	// before StartContainer(), such that the cpuset is configured before
+	// the container starts
+	RegisterContainer(p *v1.Pod, c *v1.Container, containerID string) error
+
+	// UnregisterContainer is called near UnregisterContainer() so that the manager
+	// stops trying to update that container in the reconcilation loop and
+	// any CPUs dedicated to the container are freed to the shared pool.
+	UnregisterContainer(containerID string) error
+
+	State() state.Reader
+
+	IsUnderCPUPressure() bool
+
+	lifecycle.PodAdmitHandler
+}
+
+type manager struct {
+	sync.Mutex
+	policy Policy
+	state  state.State
+
+	// containerRuntime is the container runtime service interface needed
+	// to make UpdateContainerResources() calls against the containers.
+	containerRuntime runtimeService
+
+	// kletGetter provides a method for listing all the pods on the node
+	// so all the containers can be updated in the reconciliation loop.
+	kletGetter kletGetter
+
+	// podStatusProvider provides a method for obtaining pod statuses
+	// and the containerID of their containers
+	podStatusProvider status.PodStatusProvider
+
+	// getPodsFunc gets list of active pods to consider for eviction when
+	// under pressure.
+	getPodsFunc eviction.ActivePodsFunc
+
+	// killPodFunc kills a pod selected for eviction when under pressure.
+	killPodFunc eviction.KillPodFunc
+
+	// recorder records eviction events.
+	recorder record.EventRecorder
+}
+
+var _ Manager = &manager{}
+
+// NewManager creates new cpu manager based on provided policy
+func NewManager(
+	cpuPolicyName string,
+	cr runtimeService,
+	kletGetter kletGetter,
+	statusProvider status.PodStatusProvider,
+	capacityProvider eviction.CapacityProvider,
+	getPodsFunc eviction.ActivePodsFunc,
+	killPodFunc eviction.KillPodFunc,
+	recorder record.EventRecorder,
+) (Manager, lifecycle.PodAdmitHandler, error) {
+	var policy Policy
+
+	switch policyName(cpuPolicyName) {
+	case PolicyNone:
+		policy = NewNonePolicy()
+	case PolicyStatic:
+		machineInfo, err := kletGetter.GetCachedMachineInfo()
+		if err != nil {
+			return nil, nil, err
+		}
+		topo, err := topology.Discover(machineInfo)
+		if err != nil {
+			return nil, nil, err
+		}
+		glog.Infof("[cpumanager] detected CPU topology: %v", topo)
+
+		resources := capacityProvider.GetNodeAllocatableReservation()
+		cpuResource, ok := resources[v1.ResourceCPU]
```

Comment:
As mentioned in the other comment, if we do this then we avoid needing the CPU pressure condition and evictions but open the door to this situation:

* Given a system with 4 logical CPUs on Node N1.
* A user carefully creates pod P1 to obtain 4 exclusive cores.
* The scheduler sees 4 CPUs available on N1 and binds P1 there.
* The Kubelet synchronizes but is unable to run the pod due to the hard-coded shared pool low-water-mark.
* User is... baffled?

Furthermore, in a cluster where a large mix of exclusive CPU containers are present there would be no easy way to prevent this. Some fraction of guaranteed pods will fail. If I were operating such a cluster, I would work around the problem by advertising a separate resource to account exclusively-allocatable cores, which was an explicit anti-pattern the CPU manager design tries hard to avoid.

At this point the issue isn't necessarily one of technical correctness but a difference of opinion. The conflict thus far has been about which breakdown in user expectation is least bad. I think that having a pod with no CPU request get evicted from a node with the static CPU policy enabled is much more reasonable than the above scenario and matches the wider discussion that occurred during proposal review time.